L1Trigger/L1TMuonEndCapPhase2/src/Utils/CSCUtils.cc

```diff
@@ -0,0 +1,168 @@
+#include "L1Trigger/L1TMuonEndCapPhase2/interface/DAQ/TPrimitives.h"
+
+#include "L1Trigger/L1TMuonEndCapPhase2/interface/Utils/CSCUtils.h"
+
+namespace emtf::phase2::csc {
+
+  // Chambers
+  int next_10deg_chamber(int chamber) { return (chamber == 36) ? 1 : (chamber + 1); }
+
+  int prev_10deg_chamber(int chamber) { return (chamber == 1) ? 36 : (chamber - 1); }
+
+  int next_20deg_chamber(int chamber) { return (chamber == 18) ? 1 : (chamber + 1); }
+
+  int prev_20deg_chamber(int chamber) { return (chamber == 1) ? 18 : (chamber - 1); }
+
+  // Sectors
+  bool is_in_sector(int sp_endcap, int sp_sector, int tp_endcap, int tp_sector) {
+    return sp_endcap == tp_endcap && sp_sector == tp_sector;
+  }
+
+  bool is_in_neighbor_sector(
+      int sp_endcap, int sp_sector, int tp_endcap, int tp_sector, int tp_subsector, int tp_station, int tp_id) {
+    // Match endcap and neighbor sector
+    int neighbor_sector = ((sp_sector == 1) ? 6 : sp_sector - 1);
+
+    if ((sp_endcap != tp_endcap) || (neighbor_sector != tp_sector))
+      return false;
+
+    // Match CSCID in station 1
+    if (tp_station == 1)
+      return (tp_subsector == 2) && (tp_id == 3 || tp_id == 6 || tp_id == 9);
+
+    // Match CSCID in other stations
+    return tp_id == 3 || tp_id == 9;
+  }
+
+  // Use CSC trigger "CSC ID" definitions
+  // Copied from DataFormats/MuonDetId/src/CSCDetId.cc
+  int get_id(int station, int ring, int chamber) {
+    int result = 0;
+
+    if (station == 1) {
+      result = (chamber) % 3 + 1;  // 1,2,3
+
+      switch (ring) {
+        case 1:
+          break;
+        case 2:
+          result += 3;  // 4,5,6
+          break;
+        case 3:
+          result += 6;  // 7,8,9
+          break;
+        case 4:
+          break;
+      }
+    } else {
+      if (ring == 1) {
+        result = (chamber + 1) % 3 + 1;  // 1,2,3
+      } else {
+        result = (chamber + 3) % 6 + 4;  // 4,5,6,7,8,9
+      }
+    }
+
+    return result;
+  }
+
+  // Use CSC trigger sector definitions
+  // Copied from DataFormats/MuonDetId/src/CSCDetId.cc
+  int get_trigger_sector(int station, int ring, int chamber) {
+    int result = 0;
+
+    if (station > 1 && ring > 1) {
+      result = ((static_cast<unsigned>(chamber - 3) & 0x7f) / 6) + 1;  // ch 3-8->1, 9-14->2, ... 1,2 -> 6
+    } else if (station == 1) {
+      result = ((static_cast<unsigned>(chamber - 3) & 0x7f) / 6) + 1;  // ch 3-8->1, 9-14->2, ... 1,2 -> 6
+    } else {
+      result = ((static_cast<unsigned>(chamber - 2) & 0x1f) / 3) + 1;  // ch 2-4-> 1, 5-7->2, ...
```

Comment:
The hex codes are a little confusing out of context. Are these bit patterns? Would they be clearer as binary literals?

Reply:
I'm not entirely sure what these constants mean myself. They were taken from here: https://github.com/cms-sw/cmssw/blob/master/DataFormats/MuonDetId/src/CSCDetId.cc